scripts/tract_indicators.qmd

```diff
@@ -0,0 +1,850 @@
+    "B25003_002") #owner-occupied units
+  
+  #Map over all states and return 5-year ACS data for the year and variables
+  #defined above
+  acs_housing = map_dfr(
+    datasets::state.abb,
+    ~ tidycensus::get_acs(
+        state = .,
+        variables = acs_vars,
+        year = as.numeric(acs_year),
+        geography = "tract",
+        survey = "acs5",
+        output = "wide")) %>%
+    #Retain the GEOID and estimates for each queried variable
+    select(GEOID, housing_units_occupied_total = B25003_001E, housing_units_occupied_owner = B25003_002E)
+  
+  #Write the data to the local path so that it's not necessary to query the API in the future
+  write_csv(acs_housing, acs_file_path) }
+```
+
+```{r Step 4 - Clean Partial and Missing Geography Codes}
+# Join national file and counties, flag and create dummy codes for missing tracts/counties
+# Note that this takes approximately 1-2 minutes with 2021 data on a laptop
+print(paste("Step 4: Merge national file and counties, flag and create dummy codes for missing tracts/counties:", Sys.time()))
+
+hmda_geo <- hmda_raw %>% 
+  mutate(
+    # Derive county_code from census_tract when the first is missing and the second is not
+    county_code = case_when(
+      is.na(county_code) & !is.na(census_tract) ~ str_sub(as.character(census_tract), 1, 5),
+      T ~ county_code),
+    # Create flag variables for when the tract or county codes are missing
+    missing_tract = case_when(is.na(census_tract) ~ 1, TRUE ~ 0), 
+    missing_county = case_when(is.na(county_code) ~ 1, TRUE ~ 0),
+    geo2010 = case_when(
+      # If Census tract is valid then geo2010 is the input census tract
+      missing_tract == 0 ~ as.character(census_tract),
+      # If the tract is missing but the county is valid then code as state + county + XXXXXX
+      missing_tract == 1 & missing_county == 0 ~ as.character(paste0(county_code, "XXXXXX")),
+      # If the tract and county are missing then code as state + XXXXXXXXX
+      missing_tract == 1 & missing_county == 1 & !is.na(state_code) ~ as.character(paste0(state_code, "XXXXXXXXX")),
+      # If the geography is completely invalid, use XXXXXXXXXXX
+      TRUE ~ "XXXXXXXXXXX"),
+    ## if the tract is in Guam, the PR, or the USVI, replace with pseudo-code
+    ## because these aren't contained in the tract-subcounty crosswalk (but we want to 
+    ## retain them)
+    geo2010 = case_when(
+      state_code %in% c("66", "72", "78") ~ paste0(county_code, "XXXXXX"),
+      T ~ geo2010))
+
+# rm(hmda_raw)
+```
+
+```{r Step 4.5 - Check Income Limits Join}
+## Some tracts don't join because of changes over time in tract boundaries
+## or because the tract codes are incorrect in the HMDA data. 
+## If, however, many are missing income limits data, the process of joining 
+## income limits hasn't gone correctly. 
+tracts_missing_income_limits = hmda_geo %>%
+  left_join(tract_level_income_limits, by = c("geo2010" = "census_tract")) %>%
```

Comment:
```
TEST_tracts_missing_income_limits <- hmda_geo %>%
      anti_join(tract_level_income_limits, by = c("geo2010" = "census_tract"))
```
`hmda_geo` has 11,483,889 observations
411,393 of those observations are not in `tract_level_income_limits`

Reply:
Thanks for checking this! It looks like most of these either are missing a census tract code or are in Guam / PR / USVI. There are 11 rows with a non-missing census tract code in one of the states we produce estimates for, but they're all bogus census tract codes I believe (e.g., "990000").